documentation/src/main/asciidoc/user_guide/query.adoc

```diff
@@ -1,24 +1,75 @@
+capable search engine in order to cover a wide range of data retrieval use cases.
 
-=== Simple example
-We're going to store _Book_ instances in Infinispan; each _Book_ will be defined as in the following example; we have to choose which properties are indexed, and for each property we can optionally choose advanced indexing options using the annotations defined in the Hibernate Search project.
+Indexing configuration relies on a schema definition, and for that Infinispan can use annotated Java classes when in library mode,
+and protobuf schemas for remote clients written in other languages. By standardizing on protobuf, Infinispan allows full interoperability
+between Java and non-Java clients.
+
+Apart from indexed queries, Infinispan can run queries over non-indexed data (<<query.indexless, indexless queries>>) and over partially indexed data (<<query.hybrid,hybrid queries>>).
+
+In terms of Search APIs, Infinispan has its own query language called _<<query.ickle,Ickle>>_, which is a subset of JP-QL providing extensions for full-text
+querying. The <<query.dsl,Infinispan Query DSL>> can be used for both embedded and remote java clients when full-text is not required; for Java embedded clients
+Infinispan offers the <<query.hibernatesearch,Hibernate Search Query API>> which supports running Lucene queries in the grid, apart from advanced search capabilities
+like Faceted and Spatial search.
+
+Finally, Infinispan has support for <<query.continuous,Continuous Queries>>, which works in a inverse manner than the other APIs: instead of creating, executing a query
+and obtain results, it allows a client to register queries that will be evaluated continuously as data in the cluster changes, generating notifications
+whenever the changed data matches the queries.
+
+=== Embedded Querying [[query.library]]
+
+Embedded querying is available when Infinispan is used as a library. No protobuf mapping is required, and both indexing and searching
+are done on top of Java objects.
+
+====  Quick example
+
+We're going to store _Book_ instances in an Infinispan cache called "books". _Book_ instances will be indexed, so we enable indexing for the cache,
+letting Infinispan <<query.autoconfig,configure the indexing automatically>>:
+
+Declarative:
+
+[source,xml]
+----
+<infinispan>
+   <cache-container default-cache="default">
+      <distributed-cache name="books">
+        <indexing index="LOCAL" auto-config="true"/>
+      </distributed-cache>
+   </cache-container>
+</infinispan>
+
+----
+
+Programmatic:
+
+[source,java]
+----
+import org.infinispan.Cache;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.configuration.cache.Index;
+import org.infinispan.configuration.global.GlobalConfigurationBuilder;
+import org.infinispan.manager.DefaultCacheManager;
+
+GlobalConfigurationBuilder gcb = GlobalConfigurationBuilder.defaultClusteredBuilder();
```

Comment:
This programmatic snippet is self contained; it is not related to the configuration above. I'll clarify it.